dapr_agents/agents/durableagent/agent.py

```diff
@@ -900,48 +919,50 @@ def _construct_messages_with_instance_history(
         else:
             instance_messages = []
 
-        # Always include long-term memory (chat_history) for context
-        # This ensures agents have access to broadcast messages and persistent context
-        long_term_memory_data = self.state.get("chat_history", [])
+        # Get messages from persistent memory (session-based, cross-workflow)
+        persistent_memory_messages = []
+        try:
+            persistent_memory_messages = self.memory.get_messages()
+            logger.info(
+                f"Retrieved {len(persistent_memory_messages)} messages for session {self.memory.session_id}"
+            )
+        except Exception as e:
+            logger.warning(f"Failed to retrieve persistent memory: {e}")
 
-        # Convert long-term memory to dict format for LLM consumption
+        # Get long-term memory from workflow state (for broadcast messages and persistent context)
+        long_term_memory_data = self.state.get("chat_history", [])
         long_term_memory_messages = []
         for msg in long_term_memory_data:
             if isinstance(msg, dict):
                 long_term_memory_messages.append(msg)
             elif hasattr(msg, "model_dump"):
                 long_term_memory_messages.append(msg.model_dump())
 
-        # For broadcast-triggered workflows, also include additional context memory
-        source = instance_data.get("source") if instance_data else None
-        additional_context_messages = []
-        if source and source != "direct":
-            # Include additional context memory for broadcast-triggered workflows
-            context_memory_data = self.memory.get_messages()
-            for msg in context_memory_data:
-                if isinstance(msg, dict):
-                    additional_context_messages.append(msg)
-                elif hasattr(msg, "model_dump"):
-                    additional_context_messages.append(msg.model_dump())
-
-        # Build chat history with:
-        # 1. Long-term memory (persistent context, broadcast messages)
-        # 2. Short-term instance messages (current workflow specific)
-        # 3. Additional context memory (for broadcast-triggered workflows)
+        # Build chat history with proper context and order
         chat_history = []
 
-        # Add long-term memory first (broadcast messages, persistent context)
-        chat_history.extend(long_term_memory_messages)
+        # First add persistent memory and long-term memory as user messages for context
+        # This ensures we have cross-workflow context but doesn't interfere with tool state order
+        for msg in persistent_memory_messages + long_term_memory_messages:
+            msg_dict = msg.model_dump() if hasattr(msg, "model_dump") else dict(msg)
+            if msg_dict in chat_history:
+                continue
+            # Convert tool-related messages to user messages to avoid conversation order issues
+            if msg_dict.get("role") in ["tool", "assistant"] and (
```

Comment:
let's create an issue/todo to properly design session and storing session, as we should not need this if the session is correctly store, we should be able to just retrieve and use without sorting it out